### ISSUE TITLE
Confusion where the reference model is updated

### ISSUE THREAD

**Issue by Nusob888** [LungCellAtlas/HLCA_reproducibility#1]
Hi, 

Congratulations on the excellent work!

I was hoping for some help navigating how the reference extension was performed. The code all makes sense until code chunk 30. 
Leading up to this, I can see that models were trained and saved independently for every batch in the query_adata. Here the reference model is used to instantiate the new models. 

Then from line 30 is becomes a bit unclear. 
```
# concatenate reference and query data
adata_full = reference_adata.concatenate(query_adata, index_unique=None) #, batch_key=None, batch_categories=None)
# run through model
full_latent = sc.AnnData(model.get_latent_representation(adata=adata_full))
```

Here the reference_adata is concatenated with the query. However, it is not clear which model is used to infer the latent representation of the concatenated adata?

There is no **unified** model in the code prior to this point (just query batch specific models) and, as far as I know, scanVI and other vaes within scvi-tools cannot infer latent representations of data it has not seen before. 

Can you help me understand where this model for "adata_full" came from?

**Comment by LisaSikkema** [comment]
Hi Nusob888,

Thanks for your interest :) This is actually the principle of [scArches](https://www.nature.com/articles/s41587-021-01001-7), it can add new data to an existing modeled embedding through "architectural surgery". I'd recommend you take a look at the paper. In short, the idea is that we use an existing model (the one trained on the HLCA core, in our case) and slightly update it to enable the modeling of new data in such a way that the updated model weights *do not* affect the way the reference is embedded. I.e. each new dataset has a slightly altered version of the reference model, with the alterations only affecting the way that new dataset is embedded, while leaving the original model intact.

**Comment by Nusob888** [comment]
Hi, thanks for responding. 
Yes, I enjoyed the paper, hence testing the method out. 

Maybe my initial question wasn't so clear. I understand the concept of updating the model without affecting the weights. The confusion I have is more with the reproducibility code and the concept of "atlas extension" in relation to the way [2_scArches_mapping.ipynb](https://github.com/LungCellAtlas/HLCA_reproducibility/blob/main/notebooks/3_atlas_extension/2_scArches_mapping.ipynb) is coded. 

Particularly this applies to how the atlas is extended:
### 1. Here is how the query is iteratively trained via scarches
```

for condition in query_conditions[::-1]: # (going from small to large dataset, to see quickly if code is working)
    print(condition)
    query_subadata = query_adata[query_adata.obs.dataset == condition,:].copy()
    print(query_subadata.shape)
    # load model and set relevant variables:
    model = sca.models.SCANVI.load_query_data(
        query_subadata,
        dir_reference_model,
        freeze_dropout = True,
    )
    model._unlabeled_indices = np.arange(query_subadata.n_obs)
    model._labeled_indices = []
    print("Labelled Indices: ", len(model._labeled_indices))
    print("Unlabelled Indices: ", len(model._unlabeled_indices))
    # now train surgery model using reference model and query adata
    model.train(
        n_epochs_semisupervised=surgery_epochs,
        train_base_model=False,
        semisupervised_trainer_kwargs=dict(metrics_to_monitor=["accuracy", "elbo"], 
                                           weight_decay=0,
                                           early_stopping_kwargs=early_stopping_kwargs_surgery
                                          ),
        frequency=1
    )
    surgery_path = os.path.join(dir_surgery_models, condition)
    if not os.path.exists(surgery_path):
        os.makedirs(surgery_path)
    model.save(surgery_path, overwrite=True)
```

As I understand it, this is iteratively working through each condition batch and training a new query model independently per batch. Therefore, the result is a model per batch and saved as a **separate** model. 


Then later, the query_adata and reference_adata are concatenated. 
```
# concatenate reference and query data
adata_full = reference_adata.concatenate(query_adata, index_unique=None) #, batch_key=None, batch_categories=None)
# run through model
full_latent = sc.AnnData(model.get_latent_representation(adata=adata_full))
# copy indices (order is retained)
full_latent.obs.index = adata_full.obs.index
# copy cell type labels
full_latent.obs['cell_type'] = adata_full.obs[cell_type_key].tolist()
# copy batch labels
full_latent.obs['batch'] = adata_full.obs[condition_key].tolist()
```
Here is where it becomes confusing:

- After concatenation, model.get_latent_representation() is called to generate a latent space for the merged reference AND query. 
- However, this would not be possible. We do not have a model that has been trained on both the reference and **WHOLE** query_adata. We have only iteratively trained each batch within the query. 

The only way I can think of where you could do this, is if you trained the whole query_adata as a single anndata with batch variables assigned. But this is not what the notebook shows. 

Can you explain how the full_latent here is generated then?


### 2. Conceptual question
This then leads onto another question which I think you were starting to touch on with your reply. 

If I were to update a model with a query dataset, and I wanted to use this "updated" model as the new atlas; since this is the selling point of extending an atlas. I presume I would then train future query datasets on the updated model rather than the original core model?

Because this is not what appears to happen in your notebooks. 
For the notebook [LCA_scArches_mapping_new_data_to_hlca.ipynb](https://github.com/LungCellAtlas/mapping_data_to_the_HLCA/blob/main/notebooks/LCA_scArches_mapping_new_data_to_hlca.ipynb), you load the reference model as:

```
dir_ref_model = "../data/HLCA_reference_model"
```
But there is no code in the repo that generates this model.

Have you maybe missed some code somewhere? The logical place where this model is saved would be in [2_scArches_mapping.ipynb](https://github.com/LungCellAtlas/HLCA_reproducibility/blob/main/notebooks/3_atlas_extension/2_scArches_mapping.ipynb). However, as addressed in part 1 above, I cannot see how this model was generated.

**Comment by LisaSikkema** [comment]
Oh wow I see your point now, looks like I loaded the reference model in the cell above and used that for all query datasets, instead of re-loading the right model for every query dataset, which means my query embedding is actually not using the scArches mapping atm it seems! I should have used the embeddings calculated in the two cells above. That's a really good catch and I will need to re-run the downstream analyses. We're working on the revisions now and needed to re-run most of the code already, I will make sure to correct that and will see what comes out, it might actually improve the outcome... Thanks a lot for catching this bug. 
I'm surprised actually that scArches doesn't spit out an error in this case, @M0hammadL would that be something to include in next versions of scArches (if you guys haven't already)? How does scArches know which batch it's dealing with if it doesn't throw an error when it encounters a batch name that's not part of the model?

As to your conceptual question + questions about where the reference model is stored:
The notebook that generates the reference model is the [scANVI notebook](https://github.com/LungCellAtlas/HLCA_reproducibility/blob/main/notebooks/1_building_and_annotating_the_atlas_core/05_scANVI_integration.ipynb) in folder 1, (`dir_output_model` at the bottom there should be `dir_out`, I did some re-organizing of paths and folder structure before uploading to GitHub that wasn't perfectly updated in every notebook). The output of the scANVI notebook is the reference model that is used in the scArches notebook. That same model is used in the [mapping_data_to_the_HLCA](https://github.com/LungCellAtlas/mapping_data_to_the_HLCA) repo.

Then the more conceptual part of your question (hope I understood it correctly): it isn't really possible to use a mapping-based model as a new reference model for other query datasets, as we have this one-hot-encoded batch vector as conditional, plus the constraint that the weights of the reference model are not updated during query mapping. If we want to use the "updated" atlas as the new reference, we need to do a new iteration of "de novo" integration (e.g. using scANVI), set that as our reference and then map new query datasets to that new reference. It is therefore an iterative process of reference building, query mapping, reference updating, etc.

Thanks again for catching that bug, I will very much make sure to correct it.

**Comment by Nusob888** [comment]
I am glad it was useful :) 

It is very odd that scarches would not throw up an error; its models are pytorch based from what I recall. 

Thank you for taking the time to look into this and to be so helpful with the responses, this open dialogue is refreshing and has helped me orientate a lot. 

## Just on the note of the conceptual aspect again:

The quote from the scarches paper states 

_"These new input nodes with trainable weights are the aforementioned adaptors. Importantly, adaptors are shareable, allowing users to further customize shared reference models by downloading a reference atlas, choosing a set of available adaptors for that reference and finally incorporating the user’s own data by training query adaptors "_

Also from the [pancreas reproducibility](https://github.com/theislab/scArches-reproducibility/blob/master/notebooks/sample_runs/IterativeSurgery_with_OutOfSample/OoS%2BIS_Pancreas_Alpha%2BGamma.ipynb) notebook of the original scarches paper, their code suggests that each updated model is used to instantiate the new model. Below is an example snippet that shows this:

```
## First step in the notebook. Here the reference network is used to create the new model for target_adata_1

new_network = sca.operate(network, 
                          new_task_name='after',
                          new_conditions=target_conditions[0],
                          init='Xavier',
                          version='scArches',
                          remove_dropout=True,
                          )

target_adata_1 = target_adata[target_adata.obs[condition_key] == target_conditions[0]]

## new update_1 model is trained. 

new_network.train(target_adata_1,
                  train_size=0.8, 
                  condition_key=condition_key,
                  n_epochs=10000,
                  batch_size=128, 
                  early_stop_limit=50,
                  lr_reducer=40, 
                  save=True,
                  retrain=False,
                  )

## Then the second consecutive model is instantiated with the previous new_network for target_adata_2
## Note the previous new_network is used as the reference model 

new_network = sca.operate(new_network,
                          new_task_name='after2',
                          new_conditions=target_conditions[1],
                          init='Xavier',
                          version='scArches',
                         )

target_adata_2 = target_adata[target_adata.obs[condition_key] == target_conditions[1]]
target_adata_2

## Then target_adata_2 is trained. 

new_network.train(target_adata_2,
                  train_size=0.8, 
                  condition_key=condition_key,
                  n_epochs=10000,
                  batch_size=128, 
                  early_stop_limit=50,
                  lr_reducer=40, 
                  save=True,
                  retrain=False,
                  verbose=5)

```

Now this makes more sense to me. For me, the big appeal of scarches was that you can "live" update your atlas and model to query new unseen data. As their benchmarking shows, when you hide alpha and gamma cells from target_adata_1 and target_adata_2, the model can still embed these in distinct latent representations that preserve biological signal. 

For me, iteratively training each batch independently would defeat this goal. As each new query model is its own update of the reference and therefore do not interact with each other. In this scenario, the risk is that some batch effects are not negated. 

An example might be, if I have query_1 which has unseen alpha cells and query_2 which is a different technology with unseen alpha cells. When I update them independently, there may be a risk that the technology batch effects do not align alpha cells perfectly. Whereas if I trained on query_1 first, the atlas has now learned the embedding for alpha cells and can then integrate query_2 using query_1 weights. 

Maybe @M0hammadL can give their insight into this too?

If my interpretation is right, I wonder if your HCLA can be even more improved by iteratively **updating** the model rather than independently generating models and then concatenating the latent spaces that are anchored by the reference model weights. 

Would love to be updated on any improvements if you decide to try this approach! (if @M0hammadL confirms to be a valid approach)

**Comment by M0hammadL** [comment]
hi @LisaSikkema @Nusob888  I just had a look at your discussions, interesting conversations :) a few notes on my side: 

- The query datasets and model have to be same and there should be one model updated sequentially which seems to be not the case. Overall the quality of the integration heavily relies on the initial reference size and diversity in terms unique cell-types and number of batches, and the bigger and more diverse the better query data would be integrated 
- 
- @Nusob888  those snippets have you pasted above (assuming it comes from reproducibility ) are a tad old from very early versions. I am not sure what happen here that the models did not throw and error since they are basically scvi-tools codebase and not from us, one-pot possibility is that the batch labels are 1-2 (same as all pair ref-qeury) models trained and thus they will go through. but we can check this

**Comment by Nusob888** [comment]
Thanks @M0hammadL. 

Just so I am absolutely clear. The core model should be updated sequentially and not independently correct?
e.g. 

```
vae

vae_1(vae)

vae_1 = scvi.model.TOTALVI.load_query_data(
  query_1, 
  vae,
  freeze_dropout = True,
 )

vae_2 = scvi.model.TOTALVI.load_query_data(
  query_2, 
  vae_1,
  freeze_dropout = True,
 )

```

**Comment by LisaSikkema** [comment]
Thanks for pitching in @M0hammadL 
I'm a bit confused now, I think computationally/output-wise it shouldn't matter if you train sequentially or in parallel, right? Updating of weights from one batch does not affect modeling of another batch, as each batch has its own adapter + weights, that are completely separate from the others and not updated when a new adapter is added, right?
Then also, as we used one adapter per batch, I wouldn't expect users with new datasets to use an existing adapter to map their data to the atlas. This would be different if you consider more general factors 'batch', such as e.g. 10X version, with one adapter for 10X 3', another for 10X 5', etc. Then definitely it's more user-friendly to keep everything in one larger model and have the ability to load the entire, extended model at once.

As to my decision to train sequentially/independently, this was based on a conversation we had @M0hammadL , where you told me that when training simultaneously, query cells are randomly selected during training, hence queries/batches with more cells would dominate the training process. But that was as a side, didn't realize there was the option of sequentially training as @Nusob888  showed above.

So my main confusion is still: I would think it shouldn't matter, computationally, if you train independently or update the model sequentially, aside from usability issues. Would you agree?

**Comment by Nusob888** [comment]
Hi @LisaSikkema, 

Keen to hear what @M0hammadL has to say. 

From my perspective, the issue with the independent training of batches is that there is no consensus model at the end. 
In that sense, the extension of the atlas would be moot as the model has not been updated with new data. 

The point about the adapter is definitely interesting: I understood it that the adapter does get trained, although it does not alter the weights hugely for the core reference model. However, to create a unified extended atlas, the sequential training is needed to keep adding adapters to the core so that eventually you have a unified model that has been fed all extension datasets. 

Otherwise, there is no way of transfer labels of all "core" and "extension" datasets when referencing new data as the only model that can be used in that instance is the core, or core + 1 batch of query.

**Comment by M0hammadL** [comment]
oh @LisaSikkema  you did exactly the right thing, the weights are independent and do not influence each other for each batch, what I meant there was maybe to use one model and sequentially update it with different and make that "**one**"  final model public for users rather than having a new branched model for each query data. However, performance-wise sequential IMO works better due to dataset training biases.

**Comment by LisaSikkema** [comment]
Okay then we agree about that @M0hammadL :) 
@Nusob888 about your comment "although it does not alter the weights hugely for the core reference model", that should in fact  be: "it does not alter the weights *at all* for the core reference model, *nor for the other adapters*", hence the issue here is solely about useability. Independent training and "extraction" of the data embedding per query dataset, then concatenating those embeddings, should be exactly the same as sequential training and then simulatenously extracting the embedding for all query datasets from the joint model, the latter is just easier. Does that make sense?

**Comment by Nusob888** [comment]
Yes, all makes sense now. 

So I guess the main points relating to my original query are all answered with the outstanding issues of:
- The reproducibility code still doesn't quite make sense from the query latent space generation, but this is now being modified and will be released in the next iteration. 

@M0hammadL, it may be worth updating the scarches website with an example of sequential queries. This is not immediately obvious in the current documentation and although it is a usability issue, it may still result in many models being released without the ability to extend a unified model. 

Thank you both for taking the time to clarify, happy for you to close the issue if all is in hand. Can't wait to read the final paper.

**Comment by LisaSikkema** [comment]
Awesome :) I'll close this issue once I have uploaded the corrected notebook! Just for myself to keep track of things :)

**Comment by LisaSikkema** [comment]
So I just re-ran all the code with the correct embeddings (the one I calculate further up in the notebook), and strangely the results didn't change. My best guess is I never used that bottom piece of code to get the final embedding, and wrongly left it in when I cleaned up the notebooks. The only sign of that is that I wrote above the cell with the bug:
"....select dataset of interest (name and matching query adata subset)...."
which is not what I do in the cell below that comment (with bug), so maybe I removed some cells (in which I do the concatenating of the stored embeddings) during the notebook cleaning...
Unfortunately I have no history of those earlier notebooks as I didn't work from GitHub at the time, which I now know I should be doing. 
I am going to close this issue for now.
